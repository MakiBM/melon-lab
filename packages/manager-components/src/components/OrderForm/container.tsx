import { withFormik } from 'formik';
import { compose, withHandlers } from 'recompose';
import * as Yup from 'yup';
import {
  divide,
  greaterThan,
  max,
  min,
  multiply,
} from '../../utils/functionalBigNumber';
import OrderForm from './index';

const initialState = props => {
  const isMarket = props.strategy === 'Market' ? true : false;

  return {
    type: props.selectedOrderType ? props.selectedOrderType : '',
    exchange: props.selectedExchange ? props.selectedExchange : '',
    price: props.selectedOrder && isMarket ? props.selectedOrder : '',
    quantity: '',
    total: '',
  };
};

const claculateInputs = (props, field, value) => {
  const { values, info, strategy } = props;
  let maxTotal;
  let maxQuantity;

  const typeValue = field === 'type' ? value : values.type;
  const totalValue = field === 'total' ? value : values.total;
  const quantityValue = field === 'quantity' ? value : values.quantity;

  if (strategy === 'Market') {
    maxTotal =
      typeValue === 'Buy'
        ? min(info.tokens.quoteToken.balance, totalValue)
        : totalValue;
    maxQuantity =
      typeValue === 'Sell'
        ? max(info.tokens.baseToken.balance, quantityValue)
        : quantityValue;
  } else if (strategy === 'Limit') {
    maxTotal = typeValue === 'Buy' ? info.tokens.quoteToken.balance : Infinity;
    maxQuantity =
      typeValue === 'Sell' ? info.tokens.baseToken.balance : Infinity;
  }

  if (field === 'total') {
    if (greaterThan(value, maxTotal)) {
      props.setFieldValue('total', maxTotal);
    } else if (values.price) {
      const quantity = divide(value, values.price);
      if (values.quantity !== value) {
        props.setFieldValue('quantity', quantity);
      }
    }
  }

  if (field === 'quantity') {
    if (greaterThan(value, maxQuantity)) {
      props.setFieldValue('quantity', maxQuantity);
    } else if (values.price) {
      const total = multiply(value, values.price);
      if (values.total !== value) {
        props.setFieldValue('total', total);
      }
    }
  }

  if (field === 'price' && values.quantity) {
    props.setFieldValue('total', multiply(values.quantity, value));
  }
};

const withFormValidation = withFormik({
  mapPropsToValues: props => ({ ...initialState(props) }),
  validationSchema: Yup.object().shape({
    price: Yup.string().required('Price is required.'),
    quantity: Yup.string().required('Quantity is required.'),
    total: Yup.string().required('Total is required.'),
  }),
  handleSubmit: values => {
    // TODO: define handleSubmit
  },
});

const withFormHandler = compose(
  withHandlers({
    onChange: props => (values, event) => {
      props.setFieldValue(event.target.name, values.value);
      claculateInputs(props, event.target.name, values.value);
    },
  }),
);

export default compose(withFormHandler, withFormValidation)(OrderForm);
