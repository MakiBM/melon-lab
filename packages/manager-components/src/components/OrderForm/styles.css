.order-form__input {
  margin: 0 0 var(--space-m) 0;
}

.order-form__dropdown {
  margin: var(--space-m) 0;
}

.order-form__switch,
.order-form__order-info {
  margin: var(--space-m) 0;
}

.order-form__error {
  margin: var(--space-xxs) 0 0 var(--space-xs);
}
