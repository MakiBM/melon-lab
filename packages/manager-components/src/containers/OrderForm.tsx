import { withHandlers, defaultProps, withState } from 'recompose';
import { cleanNumber } from '~/utils/Validation';

const withInputValueState = withState('value', 'updateValue', '');
const withInputValueHandlers = withHandlers({
  onChange: props => event => {
    props.cleanNumber
      ? props.updateValue(cleanNumber(event, props.decimals))
      : props.updateValue(event.target.value);
  },
});

const withDefaultProps = defaultProps({
  info: {
    lastPrice: 8.125,
    bid: 8.125,
    ask: 8.125,
    balances: [
      {
        name: 'ETH-T',
        value: 8.125,
      },
      {
        name: 'MLN-T',
        value: 8.125,
      },
    ],
  },
  fields: {
    dropdown: {
      name: 'exchange',
      options: [
        { value: 'RadarRelay', name: 'Radar Relay' },
        { value: 'OasisDEX', name: 'OasisDEX' },
      ],
      label: 'Exchange Server',
    },
    switch: {
      options: [
        'ETH-T-M',
        'MLN-T-M',
      ],
      labels: [
        'Buy',
        'Sell',
      ],
    },
    inputs: {
      price: {
        label: 'Price',
        name: 'price',
        insideLabel: true,
        cleanNumber: true,
        decimals: 4,
        placeholder: '0.0000',
      },
      quantity: {
        label: 'Quantity',
        name: 'quantity',
        insideLabel: true,
        cleanNumber: true,
        decimals: 4,
        placeholder: '0.0000',
      },
      total: {
        label: 'Total',
        name: 'total',
        insideLabel: true,
        cleanNumber: true,
        decimals: 4,
        placeholder: '0.0000',
      },
    },
  },
});

export { withDefaultProps, withInputValueState, withInputValueHandlers };
