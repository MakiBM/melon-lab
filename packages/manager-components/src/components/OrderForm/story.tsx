import { storiesOf } from '@storybook/react';
import React from 'react';
import OrderForm from './index';

const initialProps = {
  baseTokenSymbol: 'ETH-T-M',
  quoteTokenSymbol: 'MLN-T-M',
  strategy: 'Limit',
  selectedOrder: false,
  info: {
    lastPrice: 0.5,
    bid: 0,
    ask: 0,
    tokens: {
      baseToken: {
        name: 'ETH-T',
        balance: 30.0,
      },
      quoteToken: {
        name: 'MLN-T',
        balance: 20.0,
      },
    },
  },
  exchanges: [
    { value: 'RadarRelay', name: 'Radar Relay' },
    { value: 'OasisDEX', name: 'OasisDEX' },
  ],
  selectedExchange: 'RadarRelay',
  selectedOrderType: 'Buy',
  decimals: 4,
};

storiesOf('Components|Order Form', module).add('Default', () => {
  return <OrderForm {...initialProps}>Order Form</OrderForm>;
});
