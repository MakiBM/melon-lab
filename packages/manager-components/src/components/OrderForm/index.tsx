import React, { StatelessComponent } from 'react';
import Button from '~/blocks/Button';
import Dropdown from '~/blocks/Dropdown';
import Form from '~/blocks/Form';
import Input from '~/blocks/Input';
import Switch from '~/blocks/Switch';
import Toggle from '~/blocks/Toggle';
import OrderInfo from '~/components/OrderInfo';

import styles from './styles.css';

interface FormValues {
  price: string;
  quantity: string;
  total: string;
  exchange: string;
  orderType: string;
  strategy: string;
}

export interface OrderFormProps {
  values: FormValues;
  handleSubmit?: any;
  handleBlur?: any;
  onChange?: React.ChangeEvent<any>;
  info?: any;
  baseTokenSymbol?: string;
  quoteTokenSymbol?: string;
  exchanges: Array<{
    name: string;
    label: string;
  }>;
  selectedOrder?: any;
  errors: any;
  touched: any;
  decimals?: number;
  type?: string;
  dataValid?: boolean;
}

export const OrderForm: StatelessComponent<OrderFormProps> = ({
  handleSubmit,
  handleBlur,
  onChange,
  info,
  baseTokenSymbol,
  quoteTokenSymbol,
  exchanges,
  selectedOrder,
  errors,
  values,
  touched,
  decimals,
  dataValid,
}) => {
  const isMarket = values.strategy === 'Market' ? true : false;
  const numberPlaceholder = (0).toFixed(decimals);

  return (
    <div className="order-form">
      <style jsx>{styles}</style>
      <Form>
        {!dataValid && <p>Trading not possible when price feed down</p>}
        <div className="order-form__toggles">
          <div className="order-form__toggle">
            <Toggle
              name="strategy"
              value="Market"
              text="Market"
              isChecked={values.strategy === 'Market'}
              onChange={onChange}
            />
          </div>
          <div className="order-form__toggle">
            <Toggle
              name="strategy"
              value="Limit"
              text="Limit"
              isChecked={values.strategy === 'Limit'}
              onChange={onChange}
            />
          </div>
        </div>

        <div className="order-form__switch">
          <Switch
            options={[baseTokenSymbol, quoteTokenSymbol]}
            labels={['Buy', 'Sell']}
            onChange={onChange}
            name="orderType"
            value={values.orderType}
            isChecked={values.orderType === 'Sell' ? true : false}
            disabled={isMarket || !dataValid}
          />
        </div>
        {/* <div className="order-form__dropdown">
          <Dropdown
            name="exchange"
            value={values.exchange}
            options={exchanges}
            label="Exchange"
            onChange={onChange}
            disabled={isMarket || !dataValid}
          />
        </div> */}
        <div className="order-form__order-info">
          <OrderInfo {...info} />
        </div>
        <div className="order-form__input">
          <Input
            value={values.price}
            disabled={isMarket || !dataValid}
            type="number"
            label="Price"
            name="price"
            insideLabel="true"
            placeholder={numberPlaceholder}
            decimals={decimals}
            onChange={onChange}
            onBlur={handleBlur}
            required={true}
            formatNumber={true}
            error={touched.price && errors.price}
          />
        </div>
        <div className="order-form__input">
          <Input
            value={values.quantity}
            type="number"
            label="Quantity"
            name="quantity"
            insideLabel="true"
            placeholder={numberPlaceholder}
            decimals={decimals}
            onChange={onChange}
            onBlur={handleBlur}
            required={true}
            formatNumber={true}
            error={touched.quantity && errors.quantity}
            disabled={(isMarket && !values.price) || !dataValid}
          />
        </div>
        <div className="order-form__input">
          <Input
            value={values.total}
            type="number"
            label="Total"
            name="total"
            insideLabel="true"
            placeholder={numberPlaceholder}
            decimals={decimals}
            onChange={onChange}
            onBlur={handleBlur}
            required={true}
            formatNumber={true}
            error={touched.total && errors.total}
            disabled={(isMarket && !values.price) || !dataValid}
          />
        </div>
        <Button
          disabled={(isMarket && !values.price) || !dataValid}
          onClick={handleSubmit}
          type="submit"
        >
          {values.orderType}
        </Button>
      </Form>
    </div>
  );
};

export default OrderForm;
