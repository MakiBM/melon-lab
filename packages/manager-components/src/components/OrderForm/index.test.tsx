import React from 'react';
import OrderForm from './index';

describe('OrderForm', () => {
  const defaultElement = <OrderForm>Default text</OrderForm>;

  it('should render correctly', () => {
    const wrapper = shallow(defaultElement);
    expect(wrapper).toMatchSnapshot();
  });

  xit('should be selectable by class "order-form"', () => {
    expect(shallow(defaultElement).is('.order-form')).toBe(true);
  });

  it('should mount in a full DOM', () => {
    expect(mount(defaultElement).find('.order-form').length).toBe(1);
  });

  xit('should render to static HTML', () => {
    expect(render(defaultElement).text()).toEqual('Default text');
  });
});
