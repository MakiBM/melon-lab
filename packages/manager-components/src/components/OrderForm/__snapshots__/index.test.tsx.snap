// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderForm should render correctly 1`] = `
<Formik
  baseTokenSymbol="ETH-T-M"
  dataValid={true}
  decimals={6}
  enableReinitialize={true}
  exchanges={
    Array [
      Object {
        "name": "Radar Relay",
        "value": "RadarRelay",
      },
      Object {
        "name": "OasisDEX",
        "value": "OasisDEX",
      },
    ]
  }
  handleSubmit={[Function]}
  info={
    Object {
      "ask": 0,
      "bid": 0,
      "lastPrice": 0.5,
      "tokens": Object {
        "baseToken": Object {
          "balance": 30,
          "name": "ETH-T",
        },
        "quoteToken": Object {
          "balance": 20,
          "name": "MLN-T",
        },
      },
    }
  }
  initialValues={
    Object {
      "exchange": "RADAR_RELAY",
      "price": "",
      "quantity": "",
      "total": "",
      "type": "buy",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  quoteTokenSymbol="MLN-T-M"
  render={[Function]}
  selectedExchange="RadarRelay"
  selectedOrderType="Buy"
  strategy="Limit"
  total=""
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={[Function]}
/>
`;
